// @flxbl-io/sfdx-process-wrapper/lib/commandExecutor/ExecuteCommand.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const child_process = require("child_process");
const sfp_logger_1 = __importStar(require("@flxbl-io/sfp-logger"));
const DefaultProcessOptions_1 = __importDefault(require("./DefaultProcessOptions"));
class ExecuteCommand {
    constructor(logger, logLevel, showProgress) {
        this.logger = logger;
        this.logLevel = logLevel;
        this.showProgress = showProgress;
    }
    execCommand(command, workingdirectory, timeout = 0, isEnableErrorFromOutputStream = false) {
        return new Promise((resolve, reject) => {
            try {
                let childProcess;
                childProcess = child_process.exec(command, {
                    cwd: workingdirectory,
                    timeout: timeout,
                    ...(0, DefaultProcessOptions_1.default)(),
                });
                // variables for collecting data written to STDOUT and STDERR
                let stdoutContents = '';
                let stderrContents = '';
                // collect data written to STDOUT into a string
                childProcess.stdout.on('data', (data) => {
                    stdoutContents += data.toString();
                    if (this.showProgress)
                        sfp_logger_1.default.log((0, sfp_logger_1.COLOR_TRACE)(data), sfp_logger_1.LoggerLevel.INFO, this.logger);
                });
                // collect data written to STDERR into a string
                childProcess.stderr.on('data', (data) => {
                    stderrContents += data.toString();
                });
                childProcess.once('close', (code, signal) => {
                    if (code === 0 || (code === null && signal === 'SIGTERM')) {
                        resolve(stdoutContents);
                    }
                    else {
                        if (stderrContents && !isEnableErrorFromOutputStream)
                            reject(new Error(stderrContents));
                        else
                            reject(new Error(stdoutContents));
                    }
                });
                childProcess.once('error', (err) => {
                    if (stderrContents && !isEnableErrorFromOutputStream)
                        reject(new Error(stderrContents));
                    else
                        reject(new Error(stdoutContents));
                });
            }
            catch (error) {
                reject(error);
            }
        });
    }
}
exports.default = ExecuteCommand;
