// @flxbl-io/sfp-logger/lib/SFPLogger.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.COLOR_KEY_VALUE = exports.COLOR_KEY_MESSAGE = exports.COLOR_TIME = exports.COLOR_SUCCESS = exports.COLOR_HEADER = exports.COLOR_DEBUG = exports.COLOR_TRACE = exports.COLOR_INFO = exports.COLOR_WARNING = exports.COLOR_ERROR = exports.FileLogger = exports.VoidLogger = exports.ConsoleLogger = exports.LoggerLevel = void 0;
const fs = __importStar(require("fs-extra"));
const os_1 = require("os");
const chalk = require("chalk");
var LoggerLevel;
(function (LoggerLevel) {
    LoggerLevel[LoggerLevel["TRACE"] = 10] = "TRACE";
    LoggerLevel[LoggerLevel["DEBUG"] = 20] = "DEBUG";
    LoggerLevel[LoggerLevel["INFO"] = 30] = "INFO";
    LoggerLevel[LoggerLevel["WARN"] = 40] = "WARN";
    LoggerLevel[LoggerLevel["ERROR"] = 50] = "ERROR";
    LoggerLevel[LoggerLevel["FATAL"] = 60] = "FATAL";
    LoggerLevel[LoggerLevel["HIDE"] = 70] = "HIDE";
})(LoggerLevel = exports.LoggerLevel || (exports.LoggerLevel = {}));
class ConsoleLogger {
    constructor() {
        this.logType = 1 /* LoggerType.console */;
    }
}
exports.ConsoleLogger = ConsoleLogger;
class VoidLogger {
    constructor() {
        this.logType = 3 /* LoggerType.void */;
    }
}
exports.VoidLogger = VoidLogger;
class FileLogger {
    constructor(path) {
        this.path = path;
        this.logType = 2 /* LoggerType.file */;
    }
}
exports.FileLogger = FileLogger;
exports.COLOR_ERROR = chalk.bold.red;
exports.COLOR_WARNING = chalk.keyword('orange');
exports.COLOR_INFO = chalk.white;
exports.COLOR_TRACE = chalk.gray;
exports.COLOR_DEBUG = chalk.blue;
exports.COLOR_HEADER = chalk.yellowBright.bold;
exports.COLOR_SUCCESS = chalk.green.bold;
exports.COLOR_TIME = chalk.magentaBright;
exports.COLOR_KEY_MESSAGE = chalk.magentaBright.bold;
exports.COLOR_KEY_VALUE = chalk.black.bold.bgGreenBright;
class SFPLogger {
    static enableColor() {
        chalk.level = 2;
    }
    static disableColor() {
        chalk.level = 0;
    }
    static log(message, logLevel = LoggerLevel.INFO, logger) {
        if (SFPLogger.isLogsDisabled)
            return;
        if (logLevel == null)
            logLevel = LoggerLevel.INFO;
        if (logLevel < this.logLevel)
            return;
        //Todo: Proper fix
        if (logger && logger.logType === 1 /* LoggerType.console */) {
            logger = null; //Make it nullable, so it goes to console
        }
        if (logger) {
            if (logger.logType === 3 /* LoggerType.void */) {
                return;
            }
            else if (logger.logType === 2 /* LoggerType.file */) {
                let fileLogger = logger;
                message = message === null || message === void 0 ? void 0 : message.toString().replace(/[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g, '');
                fs.appendFileSync(fileLogger.path, message + os_1.EOL, 'utf8');
            }
        }
        else {
            switch (logLevel) {
                case LoggerLevel.TRACE:
                    console.log((0, exports.COLOR_TRACE)(message));
                    break;
                case LoggerLevel.DEBUG:
                    console.log((0, exports.COLOR_DEBUG)(message));
                    break;
                case LoggerLevel.INFO:
                    console.log(message);
                    break;
                case LoggerLevel.WARN:
                    console.log((0, exports.COLOR_WARNING)(message));
                    break;
                case LoggerLevel.ERROR:
                    console.log((0, exports.COLOR_ERROR)(message));
                    break;
            }
        }
    }
    static disableLogs() {
        SFPLogger.isLogsDisabled = true;
    }
    static printHeaderLine(header, color, logLevel, logger) {
        if (SFPLogger.isLogsDisabled)
            return;
        if (header == null)
            header = '';
        const lineLength = 90;
        const leftPadLength = Math.floor((lineLength - header.length) / 2);
        const rightPadLength = lineLength - leftPadLength - header.length;
        const line = '-'.repeat(leftPadLength) + `${header}` + '-'.repeat(rightPadLength);
        if (logger) {
            if (logger.logType === 3 /* LoggerType.void */) {
                return;
            }
            else if (logger.logType === 2 /* LoggerType.file */) {
                return;
            }
            else
                console.log(color(line));
        }
        else
            console.log(color(line));
    }
}
SFPLogger.logLevel = LoggerLevel.INFO;
SFPLogger.isLogsDisabled = false;
exports.default = SFPLogger;
