// @flxbl-io/sfdx-process-wrapper/lib/SFDXCommand.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SFDXCommand = void 0;
const ExecuteCommand_1 = __importDefault(require("./commandExecutor/ExecuteCommand"));
const sfp_logger_1 = __importStar(require("@flxbl-io/sfp-logger"));
class SFDXCommand {
    constructor(target_org, project_directory, logger, logLevel) {
        this.target_org = target_org;
        this.project_directory = project_directory;
        this.logger = logger;
        this.logLevel = logLevel;
    }
    async exec(quiet = true, timeout = 0, showProgress = false) {
        let command = this.getSFDXCommand();
        //add log level to error
        command += ' --loglevel=ERROR';
        if (quiet)
            command += ` --json`;
        command += ' ' + this.getGeneratedParams();
        sfp_logger_1.default.log('Generated Command:' + command, sfp_logger_1.LoggerLevel.TRACE, this.logger);
        let executor = new ExecuteCommand_1.default(this.logger, this.logLevel, showProgress);
        //CLI writes errors to Output Stream during errors in JSON Mode, so if quiet is true, use swap output for error
        let output = await executor.execCommand(command, this.project_directory, timeout, quiet);
        if (quiet) {
            return JSON.parse(output).result;
        }
        return output;
    }
    getGeneratedSFDXCommandWithParams() {
        let command = this.getSFDXCommand();
        command += ' ' + this.getGeneratedParams();
        return command;
    }
}
exports.SFDXCommand = SFDXCommand;
